icechunk/src/format/snapshot.rs

```diff
@@ -1,74 +1,51 @@
-use std::{collections::BTreeMap, convert::Infallible, sync::Arc};
+use std::{collections::BTreeMap, convert::Infallible, num::NonZeroU64, sync::Arc};
 
+use bytes::Bytes;
 use chrono::{DateTime, Utc};
 use err_into::ErrorInto;
 use flatbuffers::{FlatBufferBuilder, VerifierOptions};
 use itertools::Itertools as _;
 use serde::{Deserialize, Serialize};
 use serde_json::Value;
 
-use crate::metadata::{
-    ArrayShape, ChunkKeyEncoding, ChunkShape, Codec, DataType, DimensionNames, FillValue,
-    StorageTransformer, UserAttributes,
-};
-
 use super::{
     flatbuffers::gen,
     manifest::{Manifest, ManifestExtents, ManifestRef},
     AttributesId, ChunkIndices, IcechunkFormatError, IcechunkFormatErrorKind,
-    IcechunkResult, ManifestId, NodeId, Path, SnapshotId, TableOffset,
+    IcechunkResult, ManifestId, NodeId, Path, SnapshotId,
 };
 
 #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
-pub struct UserAttributesRef {
-    pub object_id: AttributesId,
-    pub location: TableOffset,
-}
-
-#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
-pub enum UserAttributesSnapshot {
-    Inline(UserAttributes),
-    Ref(UserAttributesRef),
-}
-
-#[derive(Clone, Debug, Hash, PartialEq, Eq)]
-pub enum NodeType {
-    Group,
-    Array,
+pub struct DimensionShape {
+    array_length: u64,
+    chunk_length: u64,
 }
 
-#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
-pub struct ZarrArrayMetadata {
-    pub shape: ArrayShape,
-    pub data_type: DataType,
-    pub chunk_shape: ChunkShape,
-    pub chunk_key_encoding: ChunkKeyEncoding,
-    pub fill_value: FillValue,
-    pub codecs: Vec<Codec>,
-    pub storage_transformers: Option<Vec<StorageTransformer>>,
-    pub dimension_names: Option<DimensionNames>,
+impl DimensionShape {
+    pub fn new(array_length: u64, chunk_length: NonZeroU64) -> Self {
+        Self { array_length, chunk_length: chunk_length.get() }
+    }
+    pub fn array_length(&self) -> u64 {
```

Comment:
what is the advantage of doing this instead of setting the fields to `pub`? Is it that we only allow gets and disallow sets this way?

Reply:
Yeah, i want to protect from having 0 chunk length, and yet, i don't want to handle the NonZero everywhere